### ISSUE TITLE
[HOLD for #16078][$4000] The emoji recommendation doesn’t show up on android on going back from the emoji picker unlike  Mweb chrome

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16364]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open staging dot on chrome on mweb
2. Go to any chat and type in :person to get the recommendation list. Now immediately click on the emoji icon on the chat box to open up the emoji picker (Do not pick any)
3. Next click somewhere on the screen above the emoji picker to go back
4. Notice that the keyboard is focused and the emoji recommendation still appears
5. But now follow the same process on android and see that the emoji recommendation list doesn’t appear even though the keyboard is still focused

## Expected Result:
Emoji recommendation list should appear on android like that of the mweb chrome to maintain consistency

## Actual Result:
Emoji recommendation list doesn’t appear on android when coming back from the emoji picker and is inconsistent with the mweb chrome

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.88-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/226755326-6a1cc6c3-0b58-4127-8e98-6697d9dc34f5.mp4

https://user-images.githubusercontent.com/43996225/226755352-ce000be8-f1fe-48ca-9683-a766874db4a0.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1679395061542339

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019708d74295d1af77</li>
        <li>Upwork Job ID: 1640189765954387968</li>
        <li>Last Price Increase: 2023-06-01</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by kadiealexander** [comment]
Can't reproduce in TestFlight, have asked @jliexpensify to test on a physical android device for me.

**Comment by jliexpensify** [comment]
Ok I can repro in v88, the same as @kavimuru 's 2nd video above:

![image](https://user-images.githubusercontent.com/30815269/226817588-69781a4f-bfc9-452f-8459-5ea1e2d6cbf2.png)

But I wonder if this is just an Android "feature"? Because I just tried the same process in Slack and I get the same behaviour.

**Comment by kadiealexander** [comment]
Thanks Jli!!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~019708d74295d1af77

**Comment by MelvinBot** [comment]
Current assignee @kadiealexander is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @Santhosh-Sellavel (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @chiragsalian (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by kadiealexander** [comment]
Not overdue.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

There is inconsistency behavior of emoji suggestion between native and web. On native, after closing the emoji picker, the emoji suggestion does not appear even though we have typed the emoji code previously on the composer.

### What is the root cause of that problem?

The real issue here is that, when we focus the composer with emoji code (for example :sm) on native, the suggestion never appears. If we fix that, this inconsistency issue will be solved too. When we close the emoji picker, the main composer will take the focus and should show the suggestion, but because the issue I mentioned above, the emoji suggestion never appears. Why?

Currently, we use `onSelectionChange` to decide whether we should show the emoji suggestion or not.
https://github.com/Expensify/App/blob/1d3f0695c79c0bad8d6cf7402536025ce1272a54/src/pages/home/report/ReportActionCompose.js#L260-L263
https://github.com/Expensify/App/blob/1d3f0695c79c0bad8d6cf7402536025ce1272a54/src/pages/home/report/ReportActionCompose.js#L392-L418

However, if the selection is the same, the `onSelectionChange` won't be called on native. For example, if I type `:sm`, the selection start and end is at 3. Then I open emoji picker (this will blur the composer) and close it. The composer then takes the focus back with the selection position (cursor) is the same when we blur it so the `onSelectionChange` is not called. However, this behavior is different on web. On web, it will always call `onSelectionChange` even though the selection is the same. We can easily reproduce it by just creating a simple text input.

https://user-images.githubusercontent.com/50919443/227846457-c904827d-8770-4de8-9370-4111f672832f.mp4

You can see at the video above that every time we click the text input with the same selection, it always calls the `onSelectionChange`.

### What changes do you think we should make in order to solve the problem?

There are 2 options that depends on what behavior (for all platform) we want. 

1. If we want that every time we refocus the composer (this includes that when we close the emoji picker), the emoji suggestion appears, then we should call `calculateEmojiSuggestion` on text input focus (call it inside `setIsFocused` if `shouldHighlight` is true).
https://github.com/Expensify/App/blob/1d3f0695c79c0bad8d6cf7402536025ce1272a54/src/pages/home/report/ReportActionCompose.js#L270-L272
2. Otherwise, inside `onSelectionChange` function, we only call `calculateEmojiSuggestion` if the new selection is different from current selection. (this means that if we close the emoji picker, all platform won't show the suggestion)

Or maybe we can apply both?

**Comment by kaushiktd** [comment]
**Please re-state the problem that we are trying to solve in this issue.**
The emoji recommendation doesn’t show up on android on going back from the emoji picker unlike Mweb chrome.

**What is the root cause of that problem?**
when user focus on composer it will call `setIsFocused()` but when app state changes(like keyboard or and close - active to inactive), it will call `onBlur` method which called  
`this.resetSuggestedEmojis();`
to clear all the suggested emojis, this is the main root cause of this issue.
https://github.com/Expensify/App/blob/1d3f0695c79c0bad8d6cf7402536025ce1272a54/src/pages/home/report/ReportActionCompose.js#L820

**What changes do you think we should make in order to solve the problem?**
when the app state changes, it will first focus on the composer then `calculateEmojiSuggestion()` is called to get suggested emojis based on modal picker is open or not and composer has emojis code. 
so you need to call
`this.calculateEmojiSuggestion()`
in `resetSuggestedEmojis()` on the below-mentioned file:
https://github.com/Expensify/App/blob/1d3f0695c79c0bad8d6cf7402536025ce1272a54/src/pages/home/report/ReportActionCompose.js#L386

```
/**
     * Clean data related to EmojiSuggestions
     */
    resetSuggestedEmojis() {
        this.setState({
            suggestedEmojis: [],
            shouldShowSuggestionMenu: false,
        }, () => {
            this.calculateEmojiSuggestion()
        });
    }
```

**Comment by MelvinBot** [comment]
Current assignee @kadiealexander is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @0xmiroslav (`External`)

**Comment by MelvinBot** [comment]
Current assignee @chiragsalian is eligible for the External assigner, not assigning anyone new.

**Comment by kadiealexander** [comment]
@chiragsalian any thoughts on the above?

**Comment by chiragsalian** [comment]
Sorry, I didn't realize there was a question for us. 

Personally, I'm not sure if there is much value in this issue. I could ask others too because if people feel strongly about tackling this then sure.

I do have one question, on android, after Actions performed step 5 i.e., emoji recommendation list doesn’t appear even though the keyboard is still focused, what happens if you type another letter? Does the emoji recommendation list then show up? 
If not then i think its a more severe bug that should be tackled.
If emoji recommendation list shows up then i don't feel too strongly about this issue and will ask others on slack if its worth tackling at this time.

**Comment by chiragsalian** [comment]
Cool, i've asked on slack if anyone think its valuable to address. Once they vote we'll either move forward with both issues or close them out.

**Comment by chiragsalian** [comment]
Discussed on slack and its decided as not a priority for the moment. Closing it out.

**Comment by chiragsalian** [comment]
Funny enough, our slack discussion was never resolved. But sure, let's open it up to get it resolved.

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by kadiealexander** [comment]
I think we should keep the price the same since we're still undecided on if it's even worth fixing.

**Comment by kadiealexander** [comment]
@chiragsalian do you think we should bump the price?

**Comment by kadiealexander** [comment]
Not overdue.

**Comment by kadiealexander** [comment]
Still looking for proposals!

**Comment by perunt** [comment]
Hello everyone, I wanted to give you an update regarding our plans to switch to inline suggestion in this version, as discussed in issue https://github.com/Expensify/App/issues/16078 . I am currently responsible for this transition. Unless it's a critical bug or a system crash, I propose that we put any non-urgent issues related to the current system on hold, as it will soon be replaced. This will help us save time and effort in the long run. What are your thoughts on this plan?

**Comment by kadiealexander** [comment]
Placing on hold as per @perunt's excellent suggestion above.

**Comment by kadiealexander** [comment]
Still on hold!

**Comment by kadiealexander** [comment]
The issue this was on hold for has completed, and this is still reproducible. 

https://github.com/Expensify/App/assets/59587260/baec6bfb-2eb7-4ec3-bf08-7eb0efbd0da2

**Comment by kadiealexander** [comment]
It just came off hold so lets wait a bit.

**Comment by nishancx** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Emoji recommendation doesn't show back up on android/native emoji picker is dismissed.

### What is the root cause of that problem?
Emoji recommendation are calculated and shown when composer's `onSelectionChange` is fired.
Unlike in mweb, `onSelectionChange` isn't fired in android/native when focus comes back to composer (in this case, when emoji picker is close).

### What changes do you think we should make in order to solve the problem?
`calculateEmojiSuggestion` can be fired from `setIsFocused` when`focus == true`, this will display the emoji suggestions menu on android/native.
With this approach, `calculateEmojiSuggestion` will be called twice on other platforms (like android/web) but since `calculateEmojiSuggestion` is debounced with `wait: 10`, this shouldn't be an issue and `calculateEmojiSuggestion` will be executed only once.

**Comment by kushu7** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The emoji recommendation doesn’t show up on android on going back from the emoji picker unlike Mweb chrome.
### What is the root cause of that problem?
If the selection is the same, `onSelectionChange` won't be called on native devices unlike web browsers on composer `focus`.
Its react-native behavior.
This problem is not just with `Emoji Suggestion` same is happening with `Mention suggestion`.
### What changes do you think we should make in order to solve the problem?

To solve our problem and show suggestions. we can call `onSelectionChange` function `onFocus` to keep consistency between platform.

https://github.com/Expensify/App/blob/0813e2ca1476635a8324cbd9f998424663a33128/src/pages/home/report/ReportActionCompose.js#L1087
We can update onFocus function with this.
```js
 onFocus={() =>{ 
            this.setIsFocused(true);
            const selection = this.state.selection;
            this.onSelectionChange({nativeEvent:{selection}})
        }}
```
It will call `onSelectionChange` with last selection and will display `Emoji or Mention suggestion` based on input. however `onSelectionChange` will be called twice on browsers as mentioned in upper comment but it will not create any performance degradation.

If we want to keep it with native only.
then we can use `getPlatform()`
```js
 onFocus={() =>{ 
            this.setIsFocused(true);
             const platform = getPlatform();
            const isNative = platform === CONST.PLATFORM.IOS || platform === CONST.PLATFORM.ANDROID;
            if(isNative) {
            const selection = this.state.selection;
            this.onSelectionChange({nativeEvent:{selection}})
            }
        }}
```
### What alternative solutions did you explore? (Optional)
None

**Comment by perunt** [comment]
Hey there! I originally created this feature with emoji suggestions, and now I want to bring it up again because it's going to be updated soon. The new version will have inline suggestions. I'm wondering if we should consider not showing these suggestions after refocusing the input, similar to how it works in Slack. The reason behind this is that there are currently many cases to consider, and there will likely be even more in the future. I believe it would be helpful to create a small document or list outlining how it should work. This way, we can have clear guidelines and ensure the feature functions smoothly.

**Comment by kadiealexander** [comment]
@0xmiroslav any thoughts on the above?

**Comment by kadiealexander** [comment]
Discussed in Slack with the C+ team, I prematurely took this off hold as #16078 has more PRs needed to be completed. Back on hold!

**Comment by kadiealexander** [comment]
Still on hold!

**Comment by kadiealexander** [comment]
Still on hold.

**Comment by kadiealexander** [comment]
On hold.

**Comment by kadiealexander** [comment]
Still on hold.

**Comment by kadiealexander** [comment]
Still on hold.

**Comment by kadiealexander** [comment]
Still on hold, shifting to monthly.

**Comment by kadiealexander** [comment]
Still on hold

**Comment by kadiealexander** [comment]
Still on hold

**Comment by quinthar** [comment]
I just tested this and can't reproduce on a Pixel 5.  Can anyone still reproduce?

**Comment by kaushiktd** [comment]
I've tried to change to the gradle version or react-native-vision-camera version also with possible solutions. But still getting the same error message.
![Screenshot 2023-12-26 at 2 25 24 PM](https://github.com/Expensify/App/assets/43398804/d177905c-8e73-4fd4-afa4-2846d70e7a96)
![unnamed](https://github.com/Expensify/App/assets/43398804/85346f91-edbe-4ea6-a4fb-cd06b140f022)
![Screenshot 2023-12-26 at 2 26 04 PM](https://github.com/Expensify/App/assets/43398804/8460cb97-d2fd-46a9-9e0d-3f8de89f9c54)

**Comment by kadiealexander** [comment]
Sorry, back from OOO! Have asked an android using pal (@jliexpensify) to try reproduce this.

**Comment by jliexpensify** [comment]
I'm on Pixel 3a (v1.4.22-3) and I don't think I can repro this anymore.

1. Open a chat and type :smile or :person to surface list
2. Select the Emoji picker to reveal emoji keyboard
3. Click outside picker
4. I see the list

This is different to what I was seeing [here](https://github.com/Expensify/App/issues/16364#issuecomment-1478976464)
